system/security/digisign/digisign.cpp

```diff
@@ -0,0 +1,260 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+#include "jliball.hpp"
+#ifdef _USE_OPENSSL
+#include <openssl/pem.h>
+#include <openssl/err.h>
+#endif
+#include "jencrypt.hpp"
+#include "digisign.hpp"
+
+static CriticalSection digiSignCrit;
+static CriticalSection digiVerifyCrit;
+
+#define EVP_CLEANUP(key,ctx) EVP_PKEY_free(key);       \
+                             EVP_MD_CTX_destroy(ctx);
+
+#define EVP_THROW(str) {                                            \
+                         char buff[120];                            \
+                         ERR_error_string(ERR_get_error(), buff);   \
+                         throw MakeStringException(-1, str, buff);  \
+                       }
+
+
+class CDigitalSignatureManager : implements IDigitalSignatureManager
+{
+private:
+    StringAttr   publicKeyFile;
+    StringBuffer publicKeyBuff;
+    StringAttr   privateKeyFile;
+    StringBuffer privateKeyBuff;
+    StringBuffer passphraseBuffEnc;
+    bool         signingConfigured;
+    bool         verifyingConfigured;
+
+public:
+    CDigitalSignatureManager()
+    {
+        signingConfigured = false;
+        verifyingConfigured = false;
+#ifdef _USE_OPENSSL
+        //query private key file location from environment.conf
+        const char * cert=0, * privKey=0, * passPhrase=0;
+        bool rc = queryHPCCPKIKeyFiles(&cert, &privKey, &passPhrase);
+        if (!isEmptyString(cert))
+            publicKeyFile.set(cert);
+        if (!isEmptyString(privKey))
+            privateKeyFile.set(privKey);
+        if (!isEmptyString(passPhrase))
+            passphraseBuffEnc.set(passPhrase);//MD5 encrypted passphrase
```

Comment:
I think I was wrong before - isn't the passPhrase returning from queryHPCCPKIKeyFiles() already decrypted ?

Reply:
It is already decrypted by queryHPCCPKIKeyFiles.  I think it would be better if queryHPCCPKIKeyFiles returned it encrypted, and we decrypt it as needed.  Thoughts?

Reply:
I will change queryHPCCPKIKeyFiles to return the encrypted password via separate Jira